# Exercise 02 -  OLAP Cubes - Roll Up and Drill Down

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [ ]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

### Connect to the local database where Pagila is loaded

In [1]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

## Roll-up
- Stepping up the level of aggregation to a large grouping
- e.g.`city` is summed as `country`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and country. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [10]:
%%time
%%sql
SELECT sum(factsales.sales_amount) as revenue, dimdate.day, dimmovie.rating, dimcustomer.country
FROM factsales 
JOIN dimdate ON (factsales.date_key = dimdate.date_key)
JOIN dimmovie ON (factsales.movie_key = dimmovie.movie_key)
JOIN dimcustomer ON (factsales.customer_key = dimcustomer.customer_key)
GROUP BY dimdate.day, dimmovie.rating, dimcustomer.country
ORDER BY revenue DESC
LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 5.21 ms, sys: 0 ns, total: 5.21 ms
Wall time: 34 ms


[(Decimal('169.67'), 30, 'G', 'China'),
 (Decimal('156.67'), 30, 'PG', 'India'),
 (Decimal('153.64'), 30, 'NC-17', 'India'),
 (Decimal('146.67'), 30, 'PG-13', 'China'),
 (Decimal('145.66'), 30, 'R', 'China'),
 (Decimal('143.68'), 30, 'R', 'India'),
 (Decimal('137.67'), 30, 'G', 'India'),
 (Decimal('135.75'), 18, 'NC-17', 'India'),
 (Decimal('131.72'), 30, 'PG', 'China'),
 (Decimal('128.74'), 21, 'PG-13', 'India'),
 (Decimal('121.72'), 18, 'PG-13', 'India'),
 (Decimal('119.76'), 18, 'PG', 'India'),
 (Decimal('116.72'), 30, 'PG-13', 'India'),
 (Decimal('115.77'), 21, 'NC-17', 'China'),
 (Decimal('115.75'), 21, 'R', 'India'),
 (Decimal('115.72'), 27, 'NC-17', 'India'),
 (Decimal('111.75'), 17, 'PG-13', 'China'),
 (Decimal('109.76'), 17, 'NC-17', 'United States'),
 (Decimal('108.77'), 30, 'NC-17', 'China'),
 (Decimal('101.79'), 20, 'PG-13', 'India')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>China</td>
        <td>169.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>India</td>
        <td>156.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>India</td>
        <td>153.64</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>China</td>
        <td>146.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>R</td>
        <td>China</td>
        <td>145.66</td>
    </tr>
</tbody></table></div>

## Drill-down
- Breaking up one of the dimensions to a lower level.
- e.g.`city` is broken up into  `districts`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and district. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [9]:
%%time
%%sql
SELECT sum(factsales.sales_amount) as revenue, dimdate.day, dimmovie.rating, dimcustomer.district
FROM factsales 
JOIN dimdate ON (factsales.date_key = dimdate.date_key)
JOIN dimmovie ON (factsales.movie_key = dimmovie.movie_key)
JOIN dimcustomer ON (factsales.customer_key = dimcustomer.customer_key)
GROUP BY dimdate.day, dimmovie.rating, dimcustomer.district
ORDER BY revenue DESC
LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.68 ms, sys: 99 µs, total: 4.78 ms
Wall time: 34.6 ms


[(Decimal('53.88'), 30, 'PG-13', 'Southern Tagalog'),
 (Decimal('38.93'), 30, 'G', 'Inner Mongolia'),
 (Decimal('36.93'), 30, 'G', 'Shandong'),
 (Decimal('36.92'), 30, 'NC-17', 'West Bengali'),
 (Decimal('34.95'), 17, 'PG-13', 'Shandong'),
 (Decimal('32.94'), 1, 'PG', 'California'),
 (Decimal('32.93'), 18, 'NC-17', 'So Paulo'),
 (Decimal('31.93'), 21, 'R', 'So Paulo'),
 (Decimal('31.93'), 30, 'NC-17', 'Buenos Aires'),
 (Decimal('30.94'), 30, 'PG', 'Southern Tagalog'),
 (Decimal('30.93'), 30, 'PG', 'So Paulo'),
 (Decimal('30.92'), 30, 'R', 'Buenos Aires'),
 (Decimal('29.95'), 18, 'NC-17', 'Maharashtra'),
 (Decimal('29.95'), 30, 'G', 'California'),
 (Decimal('29.94'), 21, 'PG-13', 'Uttar Pradesh'),
 (Decimal('29.93'), 10, 'R', 'Maharashtra'),
 (Decimal('29.93'), 20, 'PG', 'Shandong'),
 (Decimal('28.97'), 18, 'R', 'Sumy'),
 (Decimal('28.96'), 21, 'PG-13', 'Southern Tagalog'),
 (Decimal('28.95'), 20, 'PG-13', 'Buenos Aires')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>district</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Southern Tagalog</td>
        <td>53.88</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Inner Mongolia</td>
        <td>38.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Shandong</td>
        <td>36.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>West Bengali</td>
        <td>36.92</td>
    </tr>
    <tr>
        <td>17</td>
        <td>PG-13</td>
        <td>Shandong</td>
        <td>34.95</td>
    </tr>
</tbody></table></div>